### PR TITLE
Removes proposal on tip disconnection

### PR DESCRIPTION
**Pull request by red010b37**
Should fix this:

https://github.com/NAVCoin/navcoin-core/issues/280

### CODE REVIEW

**Comment by aguycalled** [comment]
could you elaborate what the changes do and why should they fix the issue?

**Comment by red010b37** [comment]
@aguycalled  

In the listproposals RPC command, it loops through and retrieves the proposals with
```
if(pblocktree->GetProposalIndex(vec))
    {
        BOOST_FOREACH(const CFund::CProposal& proposal, vec) {
```

When we invalidate the blocks the tips and payment requests were correctly disconnected but the proposals were never removed and when requested would still return "a list of proposals" that were no longer valid.

This code removes the proposal if the block or blocks before it are invalidated. 

Also, the exploded "if statement" is to help me try and track down the seg fault

**Comment by aguycalled** [comment]
has this patch been tested?

File: src/main.cpp [hunk 1]
```diff
@@ -3429,6 +3429,10 @@ bool static DisconnectTip(CValidationState& state, const CChainParams& chainpara
     // Update chainActive and related variables.
     UpdateTip(pindexDelete->pprev, chainparams);
 
+    std::vector<CFund::CProposal> vecProposal;
+     std::vector<pair<uint256,CFund::CProposal>> vPropsalsToUpdate;
+    CFund::CProposal propsal;
```
Comment (aguycalled) [inline src/main.cpp]: the variable name looks to have a typo

File: src/main.cpp [hunk 2]
```diff
@@ -3447,6 +3451,23 @@ bool static DisconnectTip(CValidationState& state, const CChainParams& chainpara
         }
     }
 
+    if(pblocktree->GetProposalIndex(vecProposal)){
+
+         for(unsigned int i = 0; i < vecProposal.size(); i++) {
+             propsal = vecProposal[i];
+
+             if(propsal.blockhash == pindexDelete->GetBlockHash()) {
```
Comment (aguycalled) [inline src/main.cpp:3459]: blockhash points to the block where the proposal was accepted, so this condition is only triggered if the proposal was accepted on the block to disconnect

File: src/main.cpp [hunk 3]
```diff
@@ -3447,6 +3451,23 @@ bool static DisconnectTip(CValidationState& state, const CChainParams& chainpara
...
+                 vPropsalsToUpdate.push_back(make_pair(propsal.hash, propsal));
```
Comment (aguycalled) [inline src/main.cpp]: this line does not make any change to the proposal db, as the same object without modification is inserted with the proposal hash as a key

File: src/main.cpp [hunk 4]
```diff
@@ -5581,6 +5615,7 @@ void static CheckBlockIndex(const Consensus::Params& consensusParams)
         if (pindex->nChainTx == 0) assert(pindex->nSequenceId == 0);  // nSequenceId can't be set for blocks that aren't linked
         // VALID_TRANSACTIONS is equivalent to nTx > 0 for all nodes (whether or not pruning has occurred).
         // HAVE_DATA is only equivalent to nTx > 0 (or VALID_TRANSACTIONS) if no pruning has occurred.
+        if (!fHavePruned) {
```
Comment (aguycalled) [inline src/main.cpp]: is this line duplicated?

**Review by aguycalled — CHANGES_REQUESTED** [review]

**Comment by aguycalled** [comment]
@red010b37 proposals and payment requests are removed on DisconnectBlock() under     // undo transactions in reverse order

File: src/main.cpp (hunk 1 above)
Comment (red010b37) [reply src/main.cpp]: Fixed

File: src/main.cpp (hunk 4 above)
Comment (red010b37) [reply src/main.cpp]: yea oops

**Pull request closed without merging**